src/main/scala/se/nimsa/dcm4che/streams/DicomFlows.scala

```diff
@@ -62,34 +70,167 @@ object DicomFlows {
-            if (buffer.length >= 140)
-              if (isPreamble(buffer))
-                if (isHeader(buffer.drop(132)))
-                  setValidated()
-                else
+
+            if (buffer.length >= maxBufferLength) {
+              if (isPreamble(buffer)) {
+                if (isHeader(buffer.drop(132))) {
+                  if (contexts.isDefined) {
+                    val info = DicomParsing.dicomInfo(buffer.drop(132)).get
+                    validateFileMetaInformation(buffer.drop(132), info)
+                  } else {
+                    setValidated()
+                  }
+                } else {
                   setFailed()
-              else if (isHeader(buffer))
-                setValidated()
-              else
+                }
+              } else if (isHeader(buffer)) {
+                if (contexts.isDefined) {
+                  val info = DicomParsing.dicomInfo(buffer).get
+                  validateSOPClassUID(buffer, info)
+                } else {
+                  setValidated()
+                }
+              } else {
                 setFailed()
-            else
+              }
+            } else {
               pull(in)
+            }
           }
         }
 
         override def onUpstreamFinish() = {
           if (!isValidated)
-            if (buffer.length == 132 && isPreamble(buffer))
-              setValidated()
-            else if (buffer.length >= 8 && isHeader(buffer))
-              setValidated()
-            else
-              setFailed()
+            if (contexts.isDefined) {
+              if (buffer.length >= 132 && isPreamble(buffer)) {
+                val info = DicomParsing.dicomInfo(buffer.drop(132)).get
+                validateFileMetaInformation(buffer.drop(132), info)
+              } else if (buffer.length >= 8 && isHeader(buffer)) {
+                val info = DicomParsing.dicomInfo(buffer).get
+                validateSOPClassUID(buffer, info)
+              } else {
+                setFailed()
+              }
+            } else {
+              if (buffer.length == 132 && isPreamble(buffer))
+                setValidated()
+              else if (buffer.length >= 8 && isHeader(buffer))
+                setValidated()
+              else
+                setFailed()
+            }
           completeStage()
         }
 
         def isHeader(data: ByteString) = DicomParsing.dicomInfo(data).isDefined
 
+        def readHeader(buffer: ByteString, assumeBigEndian: Boolean, explicitVR: Boolean): (Int, VR, Int, Int) = {
+          val tagVr = buffer.take(8)
+          val (tag, vr) = DicomParsing.tagVr(tagVr, assumeBigEndian, explicitVR)
+          if (vr == null)
+            (tag, vr, 8, bytesToInt(tagVr, 4, assumeBigEndian))
+          else if (explicitVR)
+            if (vr.headerLength == 8)
+              (tag, vr, 8, bytesToUShort(tagVr, 6, assumeBigEndian))
+            else {
+              (tag, vr, 12, bytesToInt(buffer, 8, assumeBigEndian))
```

Comment:
In `DicomPartFlow.readHeader`, it is ensured that there are 12 bytes available before returning this. Should this be done here too?